rpc: unify Peer for DRPC and gRPC connections

cthumuluru-crdb opened the pull request:
This commit unifies the Peer's functionality for both dRPC and gRPC connections. It achieves this by introducing an interface for connection-specific behavior and injecting the appropriate dRPC or gRPC implementations.

Fixes: none
Epic: CRDB-48923
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147328)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
rpc: export RPCConn type

RPCConn type needs access outside of the rpc package. This change
exports the type so that it can be used in other packages.

Fixes: none
Epic: CRDB-48923
Release note: none

cthumuluru-crdb committed:
kv,rpc: refactor Peer to support DRPC and gRPC conns

This change refactors Peer to support both DRPC and gRPC connections, by
injecting connection specific options like:
- Dialing a new connection
- Batch streams client construction
- Connection close notifier
- Equality check for connections

This commit removes DRPC support from nodedialer which will be
reintroduced in a follow-up commit.

Fixes: none
Epic: CRDB-48923
Release note: none
